### DISCUSSION TITLE
Looking for an efficient way of drawing a pixel-perfect line diagram at different display scalings

### DISCUSSION THREAD

**Discussion by PremekTill** [AvaloniaUI/Avalonia#17361] [Q&A]
Hello,

we are using Avalonia to draw electrical diagrams and color them based on measured voltage levels. Currently, we are using Path and EllipseGeometry elements and binding the Stroke/Fill property to the appropriate voltage state. We then create UserControls for basic components (Switch, Transformer etc.) and use these to put together the diagram. However, this comes with two major drawbacks.

First, the performance starts to drop significantly as we add more elements &ndash; once there are a few dozen or even lower hundreds, arranging each page takes a second or more. Second, the composition of elements means that (logically) connected lines are parts of different Path elements, which means that we cannot use automatic line interpolation provided by the Path (see examples below). This becomes problematic especially when using display scales above 100%, where prevously pixel-perfect connections start either pulling apart or overlapping (possibly due to rounding differences).

As such, I would like to ask &ndash; does anyone have any tips on improving Path performance (or an alternative solution without using Paths) when composing many lines into a single view? And is there some way to "merge" Path segments from different controls together or otherwise prevent them from drifting apart?

Just to illustrate the second point, consider this XAML:

```
<Canvas Height="40" Width="40">
    <Path Data="M 0 10 L 10 10 L 20 10" StrokeThickness="2" Stroke="Green" />

    <Path Data="M 0 20 L 10 20" StrokeThickness="2" Stroke="Green" />
    <Path Data="M 10 20 L 20 20" StrokeThickness="2" Stroke="Green" />
</Canvas>
```

Whereas the first line will (obviously) always be drawn in a single stroke, the second pair of lines can become disconnected when using different display scalings (e.g. 175%). Setting UseLayoutRounding to False on both the Canvas and the Paths does not help.

![correct_1](https://github.com/user-attachments/assets/b9ac1acf-9b4e-438c-b8de-8d5831417185)

Correctly drawn diagram at 100% display scale.

![defects_1](https://github.com/user-attachments/assets/4ff61479-9713-42ea-88f3-a7d12fd5859c) ![defects_2](https://github.com/user-attachments/assets/7691f613-a54c-4d3f-b44e-f98b46dc06f7)

Incorrectly drawn diagrams at 150% display scale.

**Comment by timunie** [comment]
You can inherit control and override Render to do the rendering on your own. This is sometimes much quicker but requires some math from your side.

About the scaling issue, maybe a ViewBix can be handy 🤔
